sendip.1
.TH "SENDIP" "1" "April 11, 2018" "sendip 2.6.0"

.SH "NAME"
sendip \- a tool to send arbitrary IP packets

.SH "SYNOPSIS"
.\" no hyphenation
.nh
.\" no alignment
.na
.HP
\fBsendip\fR [\-h] [\-V] [\-v] [\-D\ \fIotype\fR] [\-d\ \fIdata\fR\ |\ \-f\ \fIfile\fR] [\-L\ \fInum\fR] [\-T\ time] [\-S sock_opt]... [\-p\ \fImodule\fR]... [\fImodule_options\fR]... \fIhostname\fR
.\" re-enable hyphenation and aligment
.ad
.hy

.SH "DESCRIPTION"
.PP
SendIP is a command-line tool to send arbitrary IP packets. It has a large
number of options to specify the content of every header of a RIP, RIPng, BGP,
TCP, UDP, ICMP, or raw IPv4/IPv6 packet. It also allows any data to be added
to the packet. Checksums can be calculated automatically, but if you wish to
send out wrong checksums, that is supported too.
.PP
Modules are loaded in the order the -p options appear.  The headers from
each module are put immediately inside the headers from the previous module
in the final packet.  For example, to embed bgp inside tcp inside ipv4, do:
.RS
.P
sendip -p ipv4 -p tcp -p bgp ...
.RE
.PP
If no module is given, the \-d\ \fIdata\fR
or \-f\ \fIfile\fR option must be used to set the payload (in this case it
is assumed, that the packet is of type \fBAFINET\fR, i.e. IPv4). Otherwise
either \fBipv4\fR or \fBipv6\fR must be the first module given via CLI options.

Modules may be repeated to create multiple instances of a given header type.
For example, to create an ipip tunneled packet (ipv4 inside ipv4), do
sendip\ -p\ ipv4\ \fIouter-header-args\fR -p\ ipv4\ \fIinner-header-args\fR\ ...

In the case of repeated modules, arguments are applied to the closest matching
module in the command line.

.SH "OPTIONS"
.PP
Beside the module specific options the following general options are recognized
by \fBsendip\fR:
.TP
\-h
Print a help message and exit.
.TP
\-V
Print version and exit.
.TP
\-v
Be verbose.
.TP
\-D\ \fIotype\fR
Print a hex dump from each packet before sending it out. \fIotype\fR determines
how the byte offset gets printed. Use \fBh\fR for hexadecimal, \fBd\fR
for decimal dump starting at dump offset 14 (which makes it easier to compare
with full package dumps). If \fIotype\fR is in upper-case, dump offset 0
will be used instead.
.TP
\-L\ \fInum\fR
After all general options have been read and modules initialized, run the
related packet generator \fInum\fR times (0 means indefinitely). Each time
a new packet gets generated, all module specific options applied and the
packet finally sent. So each packet sent will be identical unless
random (\fBr\fIN\fR) or file (\fBf\fIfile\fR) option values are used.
.TP
\-T\ \fItime\fR
Wait \fItime\fR seconds before generating and sending the next packet (the
default value 0 means no waiting, run as fast as possible).
Ignored if no loop parameter was given.
.TP
\-S\ \fIsopt\fR
Uses the setsockopt() call of the OS to change the flags or values. For now the
following \fIsopt\fR are supported:
.RS
.TP
\fBb\fR
Set the \fBSO_BROADCAST\fR flag to 1 (enable/disable permission to transmit
broadcast messages).
.TP
\fBi\fR
Disable setting the \fBIP_HDRINCL\fR flag to 1 (header is included with data).
This is probably not what you want.
.TP
\fB6\fR
Disable setting the IPv6 \fBIPV6_UNICAST_HOPS\fR value (hop limit for unicast)
to the value given via \fBipv6\fR module option \fB\-6h\ \fInum\fR, or if not
given, to \fB32\fR.
.RE

.TP
\-p\ \fImodule\fR
Load the given \fImodule\fR. Can be specified multiple times.
.TP
\-f\ \fIfile\fR
Read the given \fIfile\fR and use its contents as payload.
.TP
\-d\ \fIdata\fR
Use the given \fIdata\fR as payload. The following formats are understood and
tried in the following order (first match wins):
.RS 2
.TP
\fBf\fIfile\fR
file. If \fIfile\fR exists, read the next line of the given \fIfile\fR and use
this line without the trailing newline (\en) but including leading and trailing
whitespaces as \fIdata\fR and restart evaluation.
If \fIfile\fR does not exist, \fBf\fIfile\fR is taken literally.
Because there is no loop detection, one may turn the computer running this
utility into a heating machine easily by just adding the same
value (\fBf\fIfile\fR) to the \fIfile\fR.
Also note, that the same \fIfile\fR can be used as option value multiple
times. Each such usage causes \fBsendip\fR to read the next line of
the \fIfile\fR. If the end of the file has been reached, it starts from the
beginning, i.e. with the first line of the \fIfile\fR. This works, because the
contents of it gets read in and cached as soon as it gets encountered as a
valid data/option value.
.TP
\fBr\fIN\fR
random. Generate and use \fIN\fR random bytes (N is a decimal number).
.TP
\fBz\fIN\fR
zero. Generate and \fIN\fR zero (nul) bytes. (N is a decimal number)
.TP
\fBt\fIN\fR
timestamp. Generate a timestamp (see \fBgettimeofday\fR(3C)) and return the
first \fIN\fR bytes of it (N is a decimal number). If \fIN\fR > than the number
of bytes produced by the timestamp (8 on 32bit, 16 on 64bit) the remaining
bytes will be all zero (nul) bytes.
.TP
\fB0x\fIS\fR or \fB0X\fIS\fR
hex. Ignore the first two charcaters (zero and letter X or x) and interprete
all others (\fIS\fR) as a hex representation of the data to use. Non-digits
are handled in a case-insensitive manner. If a non-hex character occurs, an
error will be raised.
.TP
\fB0\fIS\fR
octal. Ignore the first character (zero) and interprete all others (\fIS\fR)
as a octal representation of the data to use. If a non-octal character occurs,
an error will be raised.
.TP
\fIdata\fR
The \fIdata\fR string is taken literally.

.P
Note that the option values for the modules described below often get evaluated
the same way as \fIdata\fR, however, depending on the type of the option a
string might not be taken literally, but as a decimal number, IP address, CIDR
notation and so on. They can be sometimes even a combination of the formats
mentioned above, e.g. one may use
"\-p\ ipv4\ \-is\ 10.1.1.r1" to generate random 10.1.1.xx source addresses.
.RE

.P
The following terms are used for the value specifications of the module options
below:
.RS 2
\fIdata\fI
Same as desribed above for option "\-f\ \fIdata\fI".
.TP
\fIdnum\fR
A decimal number, which may start with a sequence of white-space
characters (which gets get usually ignored), followed by an optional
'+' or '\-' followed by a sequence of one or more digits from 0 to 9.
If any invalid character is found, the parser usually stops without throwing an
error and converts the string accumulated so far, which might be an empty
string, which in turn gets resolved to 0.  If a number is not in the expected
range, it results either in an error, or the spare bytes in front get silently
dropped.
.TP
\fIrnum\fR
A "real" number accepts hex (prefix \fB0x\fR or \fB0X\fR),
octal (prefix \fB0\fR) and decimal numbers (\fIdnum\fR), only.
Invalid numbers/parts gets usually handled the same way as described
for \fIdnum\fR.
.TP
\fInum\fR
Accepts usually "real" numbers (\fIrnum\fR) as well as the \fBr\fIN\fR and
\fBz\fIN\fR generators. Furthermore the \fBf\fIpath\fR can be used, which is
required to finally resolve to a \fInum\fR.
Invalid numbers/parts gets usually handled the same way as described
for \fIdnum\fR.
.TP
\fIhex\fR
A hexadecimal number consists of the \fBoptional\fR prefix \fB0x\fR or \fB0X\fR
followed by a sequence of the decimal digits and letters a (or A) to f (or F)
with values 10 to 15 respectively.
Invalid parts gets usually handled the same way as described for \fIdnum\fR.
.TP
\fIv4addr\fR
An IPv4 address can be given in dot or CIDR notation, or as a single \fInum\fR.
.RS
.P
In \fBdot notation\fB
\fInum\fB.\fInum\fB.\fInum\fB.\fInum\fR,
\fInum\fB.\fInum\fB.\fInum\fR, and
\fInum\fB.\fInum\fR are allowed (a dot (\fB.\fR) without a trailing
\fInum\fR part is equivalent to \fB.0\fR). The resulting value of each \fInum\fR
must be in the range of decimal 0..255.
.P
When four parts are specified, each part is interpreted as a byte of data and
assigned from left to right to the four bytes of an IPv4 address.
.P
When a three-part address is specified, the last part is interpreted as a
16-bit quantity and placed in the right most two bytes of the network address.
.P
When a two-part address is supplied, the last part is interpreted as a
24-bit quantity and placed in the right most three bytes of the network address.
.P
When only one part, i.e. a \fBsingle \fInum\fR is given, the value is stored
directly in the network address without any byte rearrangement.

.P
\fBCIDR notation\fR is similar to the dot notation:
\fIrnum\fB.\fIrnum\fB.\fIrnum\fB.\fIrnum\fB/\fIrnum\fR.

CHANGES
Changes between sendip-0.0.1 and sendip-0.0.2
* Now compiles under libc5 as well as glibc

Changes between sendip-0.0.2 and sendip-0.0.3
* New man page, taken from one supplied by rmartini <rmartini@iis.com.br>

Changes between sendip-0.0.3 and sendip-1.0
* Makefile no longer requires GNU make or etags
* included spec file from Delvin Upton <delvin.upton@spectria.com>
* now ship source and binary RPMs as well as source tar.gz
* now compiles on FreeBSD (and hopefully Net and Open) as well as linux
* minor bug fixes
* added option for random header fields (idea roughly by ganesh@earth.li)
* switched to 2digit versions

Changes between sendip-1.0 and sendip-1.1
* Fix bug the caused incorrect tcp checksums (smitchell@realogy.com)
* Add RIP-1 and RIP-2 support from Richard Polton <Richard.Polton@msdw.com>

Changes between sendip-1.1 and sendip-1.2
* Add IPv6 support from Antti Tuominen <ajtuomin@tml.hut.fi>
* Add (some) TCP options form Alexander Talos <at@atat.at>
* Bugfixes

Changes between sendip-1.2 and sendip-1.3
* Compile fix for *BSD
* Various other bugfixes
* Removed bogus libpcap dependancy
* Now included in debian unstable, hopefully
* Ready to be included in FreeBSD ports collection, hopefully
* Added -h option

Changes between sendip-1.3 and sendip-1.4
* Added contrib directory and wake on lan script, curtesy of
  Beat Bolli <beat.bolli@earthling.net>
* RIP default option from Richard Polton

Changes between sendip-1.4 and sendip-2.0-pre2
* RIP bugfix from Richard Polton
* Massive code tidy up
* -tr no longer affects -tfe and -tfc
* TCP header length defaults to correct even if options are supplied
* Setting IP checksum prints a warning as it might not work
* Can now specify the version field of the IP header
* -if now documented correctly
* man page autobuilds from help, help is more likely to be right
* TCP over IPV6 still sends the wrong checksum, but now it warns you first
* Enable setting of IPV6 priority
* ICMP and ICMPv6 merged as they are almost identical
* RIP no longer only accepts 25 entries
* BGP support from David Ball <db@earth.li>

Changes between sendip-2.0-pre2 and sendip-2.0
* include string.h to avoid memcpy(), strlen() and strcpy() warnings in
  ipv4.c, ipv6.c, rip.c and udp.c (Antti Tuominen <ajtuomin@tml.hut.fi>)
* ipv6 checksum fixes (Antti Tuominen)
* BGP fixes (David Ball <db@earth.li>)
* Compiles on AIX (thanks to Parag Kukde <pkukde@yahoo.com>)
* -d now supports hex, octal or binary data
* man page (built automagically from hacked help2man)
* finds modules correctly after a make install
 
Changes between sendip-2.0 and sendip-2.1
* Now FHS aware (Juan Antonio Martinez <jantonio@dit.upm.es>)
* NTP support added (requested by John Deatherage <Jdeatherage@CQOS.COM>)
* Merged ipv6_csum and icmp6csum
* Use (s)random() instead of (s)rand() (Bryan Croft <bryan@gurulabs.com>)
* Added -fPIC to CFLAGS to make it compile on hppa
* Now compiles on Solaris (thanks to Dax Kelson <dax@gurulabs.com>)
- tidied up the build for platform specific LDFLAGS
- turned off profiling
- renamed our copy of getopt to gnugetopt
- created types.h with bool, u_int*_t (solaris only), and endianness stuff
* sendip:
- searches for libraries in . as well as installed path
- man page fix (pointed out by Ambar Paul <ambar.paul@wipro.com>)
* tcp.so:
- now supports -tonum to specify arbitrary TCP options
* ipv4.so:
- supports (some) IP options requested by Fabrice MARIE <fabrice@celestix.com>
- only checks a single layer of enclosing headers for IPV4
* bgp.so:
- non-enclosure in TCP is non-fatal
- doesn't segfault on platforms where static variables and .so files don't mix
* rip.so:
- only checks a single layer of enclosing headers for UDP
* icmp.so:
- only checks a single layer of enclosing headers for IPV4/IPV6
- now compiles on platforms which care about alignment

Changes between sendip-2.1 and sendip-2.2
* spec file fixes (Calum Selkirk <cselkirk@panix.com>)
* FreeBSD compile fixes
* Linux-PPC compile fixes
* More SunOS compile fixes
* Use ?= not = in Makefile to allow FreeBSD ports to work without a hack
  (Joseph Scott <joseph.scott@owp.csus.edu>) 
* Added support for RIPng over UDP/IPv6.  Completely untested, I haven't even
  read the spec... Requested by armite <armite@163.com>
* CheckSum code tidied up
* tcp.so:

sendip.c
/** sendip.c - main program code for sendip
 * Copyright 2001 Mike Ricketts <mike@earth.li>
 * License: see LICENSE
 */

#define _SENDIP_MAIN

#include <netdb.h>

/* everything else */
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <time.h>

#ifdef __sun  /* for EVILNESS workaround */
#include "ipv4.h"
#endif /* __sun */

/* Use our own getopt to ensure consistent behaviour on all platforms */
#include "gnugetopt.h"

#include "sendip_module.h"
#include "crypto_module.h"
#include "common.h"
#include "modload.h"
#include "ipv6.h"
#include "dump.h"

/* housekeeping for loaded modules and their packet data */
typedef struct _sml {
	struct _sml *next;
	struct _sml *prev;
	sendip_module *mod;
	sendip_data *pack;
	int num_opts;		/* number of options for the module to keep in mind */
} sendip_mod_li;
static sendip_mod_li *first;
static sendip_mod_li *last;

/* sockaddr_storage struct is not defined everywhere, so here is our own
	nasty version
*/
typedef struct {
	u_int16_t ss_family;
	u_int32_t ss_align;
	char ss_padding[122];
} _sockaddr_storage;

static int num_opts = 0;

static char *progname;

static int
sendpacket(sendip_data *data, char *hostname, int af_type, bool verbose,
	char dump, const char *sockopts)
{
	_sockaddr_storage *to = malloc(sizeof(_sockaddr_storage));
	const char *p;

	/* socket stuff */
	int s;										/* socket for sending */
	int sent = -2;								/* number of bytes sent */
	bool sethdrincl = FALSE, setipv6opts = FALSE;
	const int on = 1;

	/* hostname stuff */
	struct hostent *host = NULL;				/* result of gethostbyname2 */

	/* casts for specific protocols */
	struct sockaddr_in *to4 = (struct sockaddr_in *) to;	/* IPv4 */
	struct sockaddr_in6 *to6 = (struct sockaddr_in6 *) to;	/* IPv6 */
	int tolen;

	if (to == NULL) {
		PERROR("Unable to send packet")
		return -3;
	}
	if ((host = gethostbyname2(hostname, af_type)) == NULL) {
		PERROR("gethostbyname2('%s', %d) failed", hostname, af_type)
		free(to);
		return -1;
	}
	memset(to, 0, sizeof(_sockaddr_storage));

	switch (af_type) {
	case AF_INET:
		to4->sin_family = host->h_addrtype;
		memcpy(&to4->sin_addr, host->h_addr, host->h_length);
		tolen = sizeof(struct sockaddr_in);
		sethdrincl = TRUE;
		af_type = PF_INET;
		break;
	case AF_INET6:
		to6->sin6_family = host->h_addrtype;
		memcpy(&to6->sin6_addr, host->h_addr, host->h_length);
		tolen = sizeof(struct sockaddr_in6);
		setipv6opts = TRUE;
		af_type = PF_INET6;
		break;
	default:
		return -2;
	}

	if (dump) {
		char buf[128 * BUFSIZ];
		int dtype = 0;
		switch (dump) {
			case 'd': dtype = 0; break;
			case 'D': dtype = 1; break;
			case 'h': dtype = 2; break;
			case 'H': dtype = 3; break;
		}
		bdump(data->data, data->alloc_len, buf, sizeof(buf), dtype);
		printf("Final packet data:\n%s\n", buf);
	}

	if ((s = socket(af_type, SOCK_RAW, IPPROTO_RAW)) < 0) {
		PERROR("Couldn't open RAW socket")
		free(to);
		return -1;
	}

	/* Set socket options */
	if (verbose)
		printf("Setting socket options:\n");
	for (p = sockopts == NULL ? "" : sockopts; *p != '\0'; p++) {
		switch (*p) {
		case 'b':
			if (verbose)
				printf(" SO_BROADCAST\n");
			if (setsockopt(s, SOL_SOCKET, SO_BROADCAST, &on, sizeof(on)) < 0) {
				PERROR("Couldn't setsockopt SO_BROADCAST")
				goto error;
			}
			break;
		case 'i':
			sethdrincl = FALSE;
			break;
		case '6':
			setipv6opts = FALSE;
			break;
		default:
			DWARN("Invalid socket option '%c' ignored", *p)
			break;
		}
	}
	if (sethdrincl) {
		if (verbose)
			printf(" IP_HDRINCL\n");
		if (setsockopt(s, IPPROTO_IP, IP_HDRINCL, &on, sizeof(on)) < 0) {
			PERROR("Couldn't setsockopt IP_HDRINCL")
			goto error;
		}
	}
	if (setipv6opts) {
		ipv6_header *iphdr = (ipv6_header *) data->data;
		if (verbose)
			printf(" IPV6_UNICAST_HOPS\n");
		/* Setting various IPV6 header option requires using setsockopt, as
		   in RFCs 3493, 3542, 2292. */
		if (setsockopt(s, IPPROTO_IPV6, IPV6_UNICAST_HOPS,
			&(iphdr->ip6_hlim), sizeof(iphdr->ip6_hlim)) < 0)
		{
			PERROR("Couldnt set Sock options for IPv6:Hop Limit")
			goto error;
		}
	}

	/* On Solaris, it seems that the only way to send IP options or packets
		with a faked IP header length is to:
		setsockopt(IP_OPTIONS) with the IP option data and size
		decrease the total length of the packet accordingly
		I'm sure this *shouldn't* work.  But it does.
	*/
#ifdef __sun
	if ((*((char *) (data->data)) & 0x0F) != 5) {
		ip_header *iphdr = (ip_header *) data->data;
		int optlen = iphdr->header_len * 4 - 20;

		if (verbose) {
			printf(" IP_OPTIONS\n"
				"Solaris workaround enabled for %d IP option bytes\n", optlen);
		}
		iphdr->tot_len = htons(ntohs(iphdr->tot_len) - optlen);

		if (setsockopt(s, IPPROTO_IP, IP_OPTIONS,
			(char *)(data->data) + 20, optlen))
		{
			PERROR("Couldn't setsockopt IP_OPTIONS")
			goto error;
		}
	}
#endif /* __sun */

	/* Send the packet */
	sent =
		sendto(s, (char *) data->data, data->alloc_len, 0, (void *) to, tolen);
	if (sent == data->alloc_len) {
		if (verbose)
			printf("Sent %d bytes to %s\n", sent, hostname);
	} else if (sent < 0) {
		PERROR("Packet not sent")
	} else if (verbose) {
		DWARN("Only sent %d/%d bytes to %s", sent, data->alloc_len, hostname)
	}
error:
	free(to);
	close(s);
	return sent;
}

static void print_usage(void) {
	sendip_mod_li *e;
	int i;
	char lbuf[LINE_MAX], *p;

	printf(
"\nUsage: %s [-hVv] [-D otype] [-d data] [-f datafile] [-L count] [-T time] \\"
"\n         \t[-S socket_opts] [-p module]... [module_option]... hostname\n"
"\n"
"Packet data, header fields:\n"
"  fF  .. next line from file F\n"
"  rN  .. N random bytes\n"
"  zN  .. N zero (nul) bytes\n"
"  tN  .. timestamp with zero padded tail of length N\n"
"  0xH .. data in hex\n"
"  0N  .. data in octal\n"
"  *   .. literal string, decimal number, IP addr, etc. depending on opt type\n"
"\nSocket options:\n"
"  b .. set SO_BROADCAST option\n"
"  i .. IPv4: disable IP header inclusion\n"
"  6 .. IPv6: disable IPv6 headers (IPV6_UNICAST_HOPS)\n"
"\n\n"
"Modules available at compile time:\n"
"    ipv4 ipv6 icmp tcp udp bgp rip ripng ntp\n"
"    ah dest esp frag gre hop route sctp wesp\n"
, progname);

	for (e = first; e != NULL; e = e->next) {
		if (e->num_opts == 0)
			continue;
		sendip_module *mod = e->mod;
		char *shortname = strrchr(mod->name, '/');

		if (!shortname)
			shortname = mod->name;
		else
			++shortname;
		p = shortname;
		for (i = 0; p[i] != '\0'; i++) {
			lbuf[i] = toupper(p[i]);
		}
		lbuf[i] = '\0';
		printf("\n\nArguments for module %s:\n", lbuf);
		for (i = 0; i < e->num_opts; i++) {
			const char *def =
				mod->opts[i].def == NULL ? "none" : mod->opts[i].def;
			snprintf(lbuf, LINE_MAX, "-%c%s %c", mod->optchar,
				mod->opts[i].optname, mod->opts[i].arg ? 'x' : ' ');
			printf("  %11s   %s.", lbuf, mod->opts[i].description);
			if (def[0] == '\n') {
				printf("\n  %11s   Default: %s\n", " ", def+1);
			} else {
				printf("  Default: %s\n", def);
			}
		}
	}

}

static void
unload_mods(bool freeit, int verbosity) {
	sendip_mod_li *e, *p;

	p = NULL;
	for (e = first; e != NULL; e = e->next) {
		if (p)
			free(p);
		p = e;
		if (freeit)
			free(p->pack->data);
		free(p->pack);
	}
	if (p)
		free(p);
	unload_modules(verbosity);
	fargs_destroy();
}

int
main(int argc, char **const argv) {
	int i;

	struct option *opts = NULL;
	char *sockopts = NULL;
	int longindex = 0, usage = 0;
	char rbuff[31];

	bool verbosity = FALSE;

	char *data = NULL, dump = '\0';
	int datafile = -1;
	int datalen = 0;
	char *datarg = NULL;

	sendip_module *mod;
	sendip_mod_li *e, *current_e;
	int optc;

	int num_modules = 0;

	sendip_data packet;

	int loopcount = 1;
	unsigned int delaytime = 0;
	
	num_opts = 0;	
	first = last = NULL;

	progname = strrchr(argv[0], '/');
	if (progname == NULL)
		progname = argv[0];
	else
		progname++;

	/* magic random seed that gives 4 really random octets */
	srandom(time(NULL) ^ (getpid() + (42 << 15)));

	/* First, get all the builtin options, and load the modules */
	gnuopterr = 0;
	gnuoptind = 0;
	while(gnuoptind < argc
		&& (EOF != (optc = gnugetopt(argc, argv, "-p:vD:d:hf:L:S:T:V"))))
	{
		switch (optc) {
		case 'V':
			printf("sendip v%s - see https://github.com/jelmd/sendip\n",
				VERSION);
			return 0;
		case 'S':
			sockopts = strdup(gnuoptarg);
			if (sockopts == NULL) {
				PERROR("Couldn't allocate memory for socket options")
				return 1;
			}
			break;
		case 'L':
			loopcount = atoi(gnuoptarg);
			break;
		case 'T':
			delaytime = atoi(gnuoptarg);
			break;
		case 'p':
			if ((mod = load_sendip_module(gnuoptarg, &i)) != NULL) {
				e = malloc(sizeof(sendip_mod_li));
				if (e == NULL) {
					PERROR("Unable to process option -p ...");
					return 1;
				}
				e->prev = last;
				e->next = NULL;
				e->pack = NULL;
				e->mod = mod;
				e->num_opts = i ? 0 : mod->num_opts;
				last = e;
				if (last->prev)
					last->prev->next = last;
				if (!first)
					first = last;
				num_modules++;
				num_opts += e->num_opts;
			}
			break;
		case 'v':
			verbosity = TRUE;
			break;
		case 'D':
			dump = gnuoptarg[0];
			if (dump != 'h' && dump != 'H' && dump != 'd' && dump != 'D') {
				DWARN("Unknown dump type '%c' - using 'd' instead.", dump);
				dump = 'd';
			}
			break;
		case 'd':
			if (datafile == -1) {
				char sdata[BUFSIZ];

				datarg = gnuoptarg;						/* save for regen */
				datalen = opt2val(sdata, datarg, BUFSIZ);
				data = (char *) malloc(datalen);
				if (data == NULL) {
					PERROR("Unable to process option -d ...")
					return 1;
				}
				memcpy(data, sdata, datalen);
			} else {
				ERROR("Only one -d or -f option can be given")
				usage = 1;
			}
			break;
		case 'h':
			usage = 2;
			break;
		case 'f':
			if (data == NULL) {
				datafile = open(gnuoptarg, O_RDONLY);
				if (datafile == -1) {
					PERROR("Couldn't open data file")
					WARN("No data will be included")
				} else {
					datalen = lseek(datafile, 0, SEEK_END);
					if (datalen == -1) {
						PERROR("Error reading data file: lseek()")
						WARN("No data will be included")
						datalen = 0;
					} else if (datalen == 0) {
						WARN("Data file is empty.\nNo data will be included")
					} else {
						data = mmap(NULL, datalen, PROT_READ, MAP_SHARED, datafile, 0);
						if (data == MAP_FAILED) {
							PERROR("Couldn't read data file: mmap()")
							WARN("No data will be included")
							data = NULL;
							datalen = 0;
						}
					}
				}
			} else {
				ERROR("Only one -d or -f option can be given")
				usage = 1;
			}
			break;
		case '?':
		case ':':
			/* skip any further characters in this option
				this is so that -tonop doesn't cause a -p option
			*/
			nextchar = NULL; gnuoptind++;
			break;
		}
	}

/* loop - need to start before module opt processing ... */
while (--loopcount >= 0) {

	/* Build the getopt listings */
	opts = malloc((1 + num_opts) * sizeof(struct option));
	if (opts == NULL) {
		PERROR("Unable to process module options")
		return 1;
	}
	memset(opts, 0, (1 + num_opts) * sizeof(struct option));
	i = 0;
	for (e = first; e != NULL; e = e->next) {
		mod = e->mod;
		int j;
		char *s;   // nasty kludge because option.name is const
		for (j = 0; j < e->num_opts; j++) {
			/* +2 on next line is one for the char, one for the trailing null */
			opts[i].name = s = malloc(strlen(mod->opts[j].optname) + 2);
			sprintf(s, "%c%s", mod->optchar, mod->opts[j].optname);
			opts[i].has_arg = mod->opts[j].arg;
			opts[i].flag = NULL;
			opts[i].val = mod->optchar;
			i++;
		}
	}
	if (verbosity)
		printf("%d module options found\n", num_opts);

	/* Initialize all */
	for (e = first; e != NULL; e = e->next) {
		mod = e->mod;
		if (verbosity)
			printf("Initializing module %s\n", mod->name);
		if (e->pack)
			free(e->pack);	/* may contain data when looping */
		e->pack = mod->initialize();
	}

	/* Get opt like getopt_long, but '-' as well as '--' can indicate a long
	 * option. If an option that starts with '-' (not '--') doesn't match a
	 * long option, but does match a short option, it is parsed as a short
	 * option instead.
	 *
	 * Apply options to the most recently invoked module first to allow
	 * separate arguments for multiply-invoked modules, e.g. for creating ipip
	 * tunneled packets.
	 */
	gnuopterr = 1;
	gnuoptind = 0;
	current_e = NULL;
	while(EOF != (optc =
		_getopt_internal(argc, argv, "p:vD:d:hf:L:S:T:V", opts, &longindex, 1)))
	{
		switch (optc) {
		case 'p':
			current_e = (current_e) ? current_e->next : first;
			break;
		case 'v':
		case 'D':
		case 'd':
		case 'f':
		case 'h':
		case 'L':
		case 'S':
		case 'T':
		case 'V':
			/* Processed above */
			break;
		case ':':
			usage = 1;
			DERROR("Option '%s' requires an argument", opts[longindex].name)
			break;
		case '?':
			usage = 1;
			DERROR("Option starting with '%c' not recognized\n", gnuoptopt)
			break;
		default:
			/* check current mod first */
			mod = NULL;
			if (current_e->mod->optchar == optc) {
				mod = current_e->mod;
				e = current_e;
			} else {
				for (e = first; e != NULL; e = e->next) {
					if (e->mod->optchar == optc) {
						mod = e->mod;
						break;
					}
				}
			}
			if (mod) {
				/* Random option arguments */
				if (gnuoptarg != NULL && strcmp(gnuoptarg,"r") == 0) {
					/* need a 32 bit number, but random() is signed and
					   nonnegative so only 31bits - we simply repeat one */
					unsigned long r = (unsigned long) random() << 1;
					r += (r & 0x00000040) >> 6;
					sprintf(rbuff, "%lu", r);
					gnuoptarg = rbuff;
				}

				if (!mod->do_opt(opts[longindex].name, gnuoptarg, e->pack)) {
					usage = 1;
				}
			}
			break;
		}
	}

	/* gnuoptind is the first thing that is not an option - should have exactly
	   one hostname... */
	if (usage == 0) {
		if (argc != gnuoptind + 1) {
 			usage = 1;
			if (argc - gnuoptind < 1) {
				ERROR("No hostname specified")
			} else {
				ERROR("More than one hostname specified")
			}
		} else if (first && first->mod->set_addr) {
			first->mod->set_addr(argv[gnuoptind], first->pack);
		}
	}

	/* free opts now we have finished with it */
	for (i = 0; i <= num_opts; i++) {
		if (opts[i].name != NULL)
			/* little trick to workaround const compiler warning */
			free((void *) (unsigned long) opts[i].name);
	}
	free(opts); /* don't need them any more */

	if (usage) {
		print_usage();
		unload_mods(TRUE, verbosity);
		if (datafile != -1) {
			munmap(data, datalen);
			close(datafile);
			datafile = -1;
		}
		if (datarg)
			free(data);
		free(sockopts);
		return 0;
	}


	/* EVIL EVIL EVIL! */
	/* Stick all the bits together - we allow finalize to shrink, but
	 * not expand the packet size and not to change the data location.
	 * Of course, any finalize which does so is responsible for pulling back
	 * all the later packet data into the area that will be sent.
	 *
	 * All of this is to accommodate esp, which needs to put its trailer after
	 * the packet data, with some padding for alignment. Since esp can't know
	 * how much padding will be needed until the rest of the packet is filled
	 * out, it preallocates an excess of padding first, and then trims in
	 * finalize to the amount actually needed.
	 */
	packet.data = NULL;
	packet.alloc_len = 0;
	packet.modified = 0;
	for (e = first; e != NULL; e = e->next) {
		packet.alloc_len += e->pack->alloc_len;
	}
	if (data != NULL)
		packet.alloc_len += datalen;
	packet.data = malloc(packet.alloc_len);
	for (i = 0, e = first; e != NULL; e = e->next) {
		memcpy((char *) packet.data + i, e->pack->data, e->pack->alloc_len);
		free(e->pack->data);
		e->pack->data = (char *)packet.data + i;
		i += e->pack->alloc_len;
	}

	/* Add any data */
	if (data != NULL)
		memcpy((char *) packet.data + i, data, datalen);

	/* Finalize from inside out */
	{
		char hdrs[num_modules];
		sendip_data *headers[num_modules];
		sendip_data d;

		d.alloc_len = datalen;
		d.data = (char *) packet.data + packet.alloc_len - datalen;

		for (i = 0, e = first; e != NULL; e = e->next, i++) {
			hdrs[i] = e->mod->optchar;
			headers[i] = e->pack;
		}

		for (i = num_modules - 1, e = last; e != NULL; e = e->prev, i--) {
			if (verbosity)
				printf("Finalizing module %s\n", e->mod->name);
			/* Remove this header from enclosing list but don't erase the
			 * header type, so that it's available to upper-level headers where
			 * needed. Instead, we tell the upper-level headers where they are
			 * in the list.
			 * wesp needs to see the esp header info, so we can't erase that,
			 * either. */
			/* headers[i] = NULL; */
			e->mod->finalize(hdrs, headers, i, &d, e->pack);

			/* Get everything ready for the next call */
			d.data = (char *) d.data - e->pack->alloc_len;
			d.alloc_len += e->pack->alloc_len;
		}
		/* Trim back the packet length if need be */
		if (d.alloc_len < packet.alloc_len)
			packet.alloc_len = d.alloc_len;
	}
	/* We could (and should?) free any leftover priv data here. */

	/* And send the packet */
	{
		int af_type;
		if (first == NULL) {
			if (data == NULL) {
				ERROR("Nothing specified to send!")
				print_usage();
				free(packet.data);
				unload_mods(FALSE, verbosity);
				free(sockopts);
				return 1;
			} else {
				af_type = AF_INET;
			}
		}
		else if (first->mod->optchar == 'i')
			af_type = AF_INET;
		else if (first->mod->optchar == '6')
			af_type = AF_INET6;
		else {
			ERROR("Either IPv4 or IPv6 must be the outermost packet")
			free(packet.data);
			unload_mods(FALSE, verbosity);
			free(sockopts);
			return 1;
		}
		i = sendpacket(&packet, argv[gnuoptind], af_type, verbosity, dump,
			sockopts);
		free(packet.data);
	}
	/* Regenerate data on subsequent loop calls */
	if (loopcount && datarg) {
		char sdata[BUFSIZ];
		int newlen;

		newlen = opt2val(sdata, datarg, BUFSIZ);
		if (newlen > datalen) {
			free(data);
			if ((data = (char *) malloc(datalen)) == NULL) {
				PERROR("Unable to allocate data memory")
				break;
			}
		}
		datalen = newlen;
		memcpy(data, sdata, datalen);
	}
	if (loopcount && delaytime)
		sleep(delaytime);
} /* end of loop */

	unload_mods(FALSE, verbosity);
	if (datafile != -1) {
		munmap(data, datalen);
		close(datafile);
		datafile = -1;
	}
	if (datarg)
		free(data);
	free(sockopts);

	return 0;
}

/* vim: ts=4 sw=4 filetype=c
 */
